net-tools/netwatch/src/udp.rs

```diff
@@ -51,8 +66,616 @@ impl UdpSocket {
+
+            let res = socket.try_io(Interest::WRITABLE, || state.send(socket.into(), transmit));
+
+            match res {
+                Ok(()) => return Ok(()),
+                Err(err) => match self.handle_write_error(err) {
+                    Some(err) => return Err(err),
+                    None => {
+                        continue;
+                    }
+                },
+            }
+        }
+    }
+
+    /// quinn based `poll_recv`
+    pub fn poll_recv_quinn(
+        &self,
+        cx: &mut Context,
+        bufs: &mut [std::io::IoSliceMut<'_>],
+        meta: &mut [quinn_udp::RecvMeta],
+    ) -> Poll<std::io::Result<usize>> {
+        loop {
+            if let Err(err) = self.maybe_rebind() {
+                return Poll::Ready(Err(err));
+            }
+
+            let guard = futures_lite::ready!(self.poll_read_socket(&self.recv_waker, cx));
+            let (socket, state) = guard.try_get_connected()?;
+
+            match socket.poll_recv_ready(cx) {
+                Poll::Pending => {
+                    self.recv_waker.register(cx.waker());
+                    return Poll::Pending;
+                }
+                Poll::Ready(Ok(())) => {
+                    // We are ready to read, continue
+                }
+                Poll::Ready(Err(err)) => match self.handle_read_error(err) {
+                    Some(err) => return Poll::Ready(Err(err)),
+                    None => {
+                        continue;
+                    }
+                },
+            }
+
+            let res = socket.try_io(Interest::READABLE, || state.recv(socket.into(), bufs, meta));
+            match res {
+                Ok(count) => {
+                    for meta in meta.iter().take(count) {
+                        trace!(
+                            src = %meta.addr,
+                            len = meta.len,
+                            count = meta.len / meta.stride,
+                            dst = %meta.dst_ip.map(|x| x.to_string()).unwrap_or_default(),
+                            "UDP recv"
+                        );
+                    }
+                    return Poll::Ready(Ok(count));
+                }
+                Err(err) => {
+                    // ignore spurious wakeups
+                    if err.kind() == std::io::ErrorKind::WouldBlock {
+                        continue;
+                    }
+                    match self.handle_read_error(err) {
+                        Some(err) => return Poll::Ready(Err(err)),
+                        None => {
+                            continue;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Whether transmitted datagrams might get fragmented by the IP layer
+    ///
+    /// Returns `false` on targets which employ e.g. the `IPV6_DONTFRAG` socket option.
+    pub fn may_fragment(&self) -> std::io::Result<bool> {
```

Comment:
I wonder for these 3 functions whether you could store them at construction and rebind time.  That way they could keep the signature without `Result`, which matches the signature higher up and avoids the odd `.unwrap_or_default()` there.

It is a lot of effort in order to give out real numbers for an effectively closed socket though.  So understand if you don't find it worth it or improves things much.

Reply:
it's not a bad idea, let me look at the way they actually get generated

Reply:
done